async/async.go

```diff
@@ -0,0 +1,372 @@
+		id := uuid.New().String()
+		// TODO: Probably set the CSV extension in the signed upload URL instead.
+		destPath := filepath.Join("/", "mnt", "raw_portfolios", fmt.Sprintf("%s.csv", id))
+		if err := h.downloadBlob(ctx, string(srcURI), destPath); err != nil {
+			return fmt.Errorf("failed to download raw portfolio blob: %w", err)
+		}
+	}
+
+	processedDir := filepath.Join("/", "mnt", "processed_portfolios")
+	if err := os.MkdirAll(processedDir, 0600); err != nil {
+		return fmt.Errorf("failed to create directory to download blob to: %w", err)
+	}
+
+	var stdout, stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "/usr/local/bin/Rscript", "/app/process_portfolios.R")
+	cmd.Stdout = io.MultiWriter(os.Stdout, &stdout)
+	cmd.Stderr = io.MultiWriter(os.Stderr, &stderr)
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to run process_portfolios script: %w", err)
+	}
+
+	sc := bufio.NewScanner(&stderr)
+
+	// TODO: Load from the output database file (or similar, like reading the processed_portfolios dir) instead of parsing stderr
+	var paths []string
+	for sc.Scan() {
+		line := sc.Text()
+		idx := strings.Index(line, "writing to file: " /* 17 chars */)
+		if idx == -1 {
+			continue
+		}
+		paths = append(paths, strings.TrimSpace(line[idx+17:]))
+	}
+
+	// NOTE: This code could benefit from some concurrency, but I'm opting not to prematurely optimize.
+	var out []*task.ParsePortfolioResponseItem
+	for _, p := range paths {
+		lineCount, err := countCSVLines(p)
+		if err != nil {
+			return fmt.Errorf("failed to count lines in file %q: %w", p, err)
+		}
+		fileName := filepath.Base(p)
+		blobURI := pacta.BlobURI(blob.Join(h.blob.Scheme(), destPortfolioContainer, fileName))
+		if err := h.uploadBlob(ctx, p, string(blobURI)); err != nil {
+			return fmt.Errorf("failed to copy parsed portfolio from %q to %q: %w", p, blobURI, err)
+		}
+		extension := filepath.Ext(fileName)
+		fileType, err := pacta.ParseFileType(extension)
+		if err != nil {
+			return fmt.Errorf("failed to parse file type from file name %q: %w", fileName, err)
+		}
+		out = append(out, &task.ParsePortfolioResponseItem{
+			Blob: pacta.Blob{
+				FileName: fileName,
+				FileType: fileType,
+				BlobURI:  blobURI,
+			},
+			LineCount: lineCount,
+		})
+	}
+
+	events := []publisher.Event{
+		{
+			Data: task.ParsePortfolioResponse{
+				TaskID:  taskID,
+				Request: req,
+				Outputs: out,
+			},
+			DataVersion: to.Ptr("1.0"),
+			EventType:   to.Ptr("parsed-portfolio"),
+			EventTime:   to.Ptr(time.Now()),
+			ID:          to.Ptr(string(taskID)),
+			Subject:     to.Ptr(string(taskID)),
+		},
+	}
+
+	if _, err := h.pubsub.PublishEvents(ctx, events, nil); err != nil {
+		return fmt.Errorf("failed to publish event: %w", err)
+	}
+
+	h.logger.Info("parsed portfolio", zap.String("task_id", string(taskID)))
+
+	return nil
+}
+
+// TODO(grady): Move this line counting into the image to prevent having our code do any read of the actual underlying data.
```

Comment:
Great minds - yeah lets' just remove this.

Reply:
See above

(also, not great minds, just one great mind. this is literally your comment, just moved)